File: Java/ToyLanguageGUI_exam_final/src/domain/PrgState.java
package domain;

import domain.statements.Statement;
import utils.*;
import exceptions.StmtExecException;

public class PrgState {
    private int id;
    private IExecStack<Statement> exec;
    private IOutList<Integer> messages;
    private IDictionary<String,Integer> symbolT;
    private Statement stmt;
    private IFileTable fileTable;
    private IHeap heap;
    private ILatchTable<Integer,Integer> latchTable;
    public PrgState(IExecStack<Statement> stack,IOutList<Integer> messages,IDictionary<String,Integer> symbolT,Statement stmt,IFileTable<Integer,FileData> ft,IHeap h,ILatchTable<Integer,Integer> lt){
        id = IDGenerator.generateID();
        this.exec = stack;
        this.messages = messages;
        this.symbolT = symbolT;
        this.stmt = stmt;
        this.fileTable = ft;
        this.heap = h;
        this.latchTable = lt;
    }

    public int getId(){ return id; }

    public void setId(int ID){ id = ID; }

    public IExecStack<Statement> getStack(){
        return exec;
    }

    public IOutList<Integer> getMessages(){
        return messages;
    }

    public IDictionary<String,Integer> getSymbols(){
        return symbolT;
    }

    public Statement getStatement(){
        return stmt;
    }

    public IFileTable<Integer,FileData> getFileTable() {return fileTable;}

    public IHeap<Integer,Integer> getHeap() {return heap;}

    public ILatchTable<Integer,Integer> getLatchTable(){return latchTable;}

    @Override
    public String toString(){
        StringBuffer buff = new StringBuffer();
        buff.append("\n");
        buff.append(""+"ID:"+id+"\n");
        buff.append("--------------------\n");
        buff.append("Execution stack: \n");
        buff.append("" + exec + "\n");
        buff.append("Message list: \n");
        buff.append(messages + "\n");
        buff.append("Dictionary: \n");
        buff.append(""+symbolT + "\n" );
        buff.append("Statement: \n");
        buff.append(""+stmt +"\n");
        buff.append("File table: \n");
        buff.append(""+fileTable+"\n");
        buff.append("Heap: \n");
        buff.append(""+heap+"\n");
        buff.append("Latch table: \n");
        buff.append(""+latchTable+"\n");
        return buff.toString();
    }

    public Boolean isNotCompleted(){
        if (exec.isEmpty())
            return false;
        return true;
    }

    public PrgState oneStep(){
        if(getStack().isEmpty())
            throw new StmtExecException("Stack is empty");
        Statement stmt = getStack().pop();
        return stmt.execute(this);
    }

}

File: ToyInterpretor/ToyInterpretor/Domain/PrgState.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using ToyInterpretor.Utils;
using ToyInterpretor.Domain.Statements;

namespace ToyInterpretor.Domain
{
    class PrgState
    {
        private IExecStack<Statement> exec;
        private IOutList<int> messages;
        private ISymTable<String, int> symbolT;
        private Statement stmt;
        private IFileTable<int,FileData> fileTable;
        public PrgState(IExecStack<Statement> stack, IOutList<int> messages, ISymTable<String, int> symbolT, Statement stmt,IFileTable<int,FileData> ft)
        {
            this.exec = stack;
            this.messages = messages;
            this.symbolT = symbolT;
            this.stmt = stmt;
            this.fileTable = ft;
        }

        public IExecStack<Statement> getStack()
        {
            return exec;
        }

        public IOutList<int> getMessages()
        {
            return messages;
        }

        public ISymTable<String, int> getSymbols()
        {
            return symbolT;
        }

        public Statement getStatement()
        {
            return stmt;
        }

        public IFileTable<int, FileData> getFileTable() { return fileTable; }


        
        public override String ToString()
        {
            StringBuilder buff = new StringBuilder();
            buff.Append("\n");
            buff.Append("--------------------\n");
            buff.Append("Execution stack: \n");
            buff.Append("" + exec + "\n");
            buff.Append("Message list: \n");
            buff.Append(messages + "\n");
            buff.Append("Dictionary: \n");
            buff.Append("" + symbolT + "\n");
            buff.Append("Statement: \n");
            buff.Append("" + stmt + "\n");
            buff.Append("File table: \n");
            buff.Append("" + fileTable + "\n");
       
            return buff.ToString();
        }
    }
}
